Amazon.IonObjectMapper/IonObjectSerializer.cs

```diff
@@ -25,69 +25,124 @@ public IonObjectSerializer(IonSerializer ionSerializer, IonSerializationOptions
 
         public override object Deserialize(IIonReader reader)
         {
-            var targetObject = options.ObjectFactory.Create(options, reader, targetType);
+            object targetObject = null;
+            ConstructorInfo ionConstructor = null;
+            ParameterInfo[] parameters = null;
+            object[] constructorArgs = null;
+            Dictionary<string, int> constructorArgIndexMap = null;
+
+            // Determine if we are using an annotated constructor.
+            var ionConstructors = targetType.GetConstructors(BINDINGS).Where(IsIonConstructor).Take(2);
+            if (ionConstructors.Any())
+            {
+                if (ionConstructors.Count() > 1)
+                {
+                    throw new InvalidOperationException(
+                        $"Only one constructor in class {targetType.Name} may be annotated " +
+                        "with the [IonConstructor] attribute and more than one was detected");
+                }
+
+                ionConstructor = ionConstructors.First();
+                parameters = ionConstructor.GetParameters();
+                constructorArgs = new object[parameters.Length];
+                constructorArgIndexMap = this.BuildConstructorArgIndexMap(parameters);
+            }
+            else
+            {
+                // If we are not using an annotated constructor, then we need to construct the object before stepping
+                // into the reader in case we need to read the type annotations during construction.
+                targetObject = options.ObjectFactory.Create(options, reader, targetType);
+            }
+
             reader.StepIn();
 
+            // Read the Ion and organize deserialized results into three categories:
+            // 1. Values to be set via annotated methods.
+            // 2. Properties to be set.
+            // 3. Fields to be set.
+            // Any of these deserialized results may also be used for the annotated constructor (if there is one). 
```

Comment:
Is this comment accurate? The deserialized results are used to set the values, properties, fields of the object itself, not to the annotated constructor.

Reply:
It's accurate. After the method/property/field logic we check if the current Ion field is one of the arguments for the annotated constructor and populate the `constructorArgs` array accordingly.